src/pickup.cpp

```diff
@@ -74,54 +74,119 @@ struct pickup_count {
     int count = 0;
 };
 
+static bool is_valid_auto_pickup( const item *pickup_item )
+{
+    int weight_limit = get_option<int>( "AUTO_PICKUP_WEIGHT_LIMIT" );
+    int volume_limit = get_option<int>( "AUTO_PICKUP_VOL_LIMIT" );
+
+    return pickup_item->volume() <= units::from_milliliter( volume_limit * 50 ) ||
+           pickup_item->weight() <= weight_limit * 50_gram;
+}
+
+static bool should_auto_pickup( const item *pickup_item )
+{
+    std::string item_name = pickup_item->tname( 1, false );
+    rule_state pickup_state = get_auto_pickup().check_item( item_name );
+
+    if( !is_valid_auto_pickup( pickup_item ) ) {
+        return false;
+    } else if( pickup_state == rule_state::WHITELISTED ) {
+        return true;
+    } else if( pickup_state != rule_state::BLACKLISTED ) {
+        //No prematched pickup rule found, check rules in more detail
+        get_auto_pickup().create_rule( pickup_item );
+
+        if( get_auto_pickup().check_item( item_name ) == rule_state::WHITELISTED ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static std::vector<item_location> get_pickup_list_from( item_location &container )
+{
+    item *container_item = container.get_item();
+    bool pick_all_items = true;
+
+    std::vector<item_location> pickup_list;
+    std::list<item *> contents = container_item->get_contents().all_items_top();
+    pickup_list.reserve( contents.size() );
+
+    for( item *item_to_check : contents ) {
+        if( should_auto_pickup( item_to_check ) ) {
+            // TODO: make check here for BLACKLIST
+            pickup_list.push_back( item_location( container, item_to_check ) );
+        } else if( item_to_check->is_container_empty() ) {
+            pick_all_items = false;
+        } else {
+            // get pickup list from nested item container
+            item_location location = item_location( container, item_to_check );
+            std::vector<item_location> pickup_nested = get_pickup_list_from( location );
+
+            // container with content was marked for pickup
+            if( pickup_nested.size() != 1 || pickup_nested[0] != location ) {
+                pick_all_items = false;
+            }
+            pickup_list.reserve( pickup_nested.size() + pickup_list.size() );
+            pickup_list.insert( pickup_list.end(), pickup_nested.begin(), pickup_nested.end() );
+        }
+    }
+    // all items in container were approved for pickup
+    if( !contents.empty() && pick_all_items ) {
+        bool all_batteries = true;
+        bool powered_container = container_item->ammo_capacity( ammotype( "battery" ) );
```

Comment:
I was also thinking about making a separate function for checking if an item is a battery.
Do you think it would be more appropriate to place it in `pickup.cpp` or `item.cpp`?

Reply:
Definitely `item.cpp`, it's a general item function.

Reply:
The new function has been added in https://github.com/CleverRaven/Cataclysm-DDA/compare/53766847da6ab06234824f1ae910028b7814abe6...4858b056aa09436e949e1ca499810abcd2257fc8.